File: apps/ios/mynewsapp/Networking/NewsService.swift
import Foundation

protocol NewsService {
    func search(query: String, sortBy: SortBy, pageSize: Int, fetchFullContent: Bool) async throws -> [Article]
    func topHeadlines(sortBy: SortBy, country: Country, category: Category, pageSize: Int, fetchFullContent: Bool) async throws -> [Article]
}

enum SortBy: String {
    case popularity = "popularity"
    case publishedAt = "publishedAt"
    case relevancy = "relevancy"
}

enum Country: String {
    case ae = "ae"
    case ar = "ar"
    case at = "at"
    case au = "au"
    case be = "be"
    case bg = "bg"
    case br = "br"
    case ca = "ca"
    case ch = "ch"
    case cn = "cn"
    case co = "co"
    case cu = "cu"
    case cz = "cz"
    case de = "de"
    case eg = "eg"
    case fr = "fr"
    case gb = "gb"
    case gr = "gr"
    case hk = "hk"
    case hu = "hu"
    case id = "id"
    case ie = "ie"
    case il = "il"
    case `in` = "in"
    case it = "it"
    case jp = "jp"
    case kr = "kr"
    case lt = "lt"
    case lv = "lv"
    case ma = "ma"
    case mx = "mx"
    case my = "my"
    case ng = "ng"
    case nl = "nl"
    case no = "no"
    case nz = "nz"
    case ph = "ph"
    case pl = "pl"
    case pt = "pt"
    case ro = "ro"
    case rs = "rs"
    case ru = "ru"
    case sa = "sa"
    case se = "se"
    case sg = "sg"
    case si = "si"
    case sk = "sk"
    case th = "th"
    case tr = "tr"
    case tw = "tw"
    case ua = "ua"
    case us = "us"
    case ve = "ve"
    case za = "za"
}

enum Category: String {
    case business = "business"
    case entertainment = "entertainment"
    case general = "general"
    case health = "health"
    case science = "science"
    case sports = "sports"
    case technology = "technology"
    
    var displayName: String {
        rawValue.capitalized
    }
}

File: apps/android/app/src/main/java/com/pouriaalmassi/mynewsapp/networking/NewsService.kt
package com.pouriaalmassi.mynewsapp.networking

import com.pouriaalmassi.mynewsapp.models.Article

interface NewsService {
    suspend fun search(
        query: String,
        sortBy: SortBy = SortBy.POPULARITY,
        pageSize: Int = 20,
        fetchFullContent: Boolean = false
    ): List<Article>

    suspend fun topHeadlines(
        sortBy: SortBy = SortBy.POPULARITY,
        country: Country = Country.US,
        category: Category = Category.BUSINESS,
        pageSize: Int = 20,
        fetchFullContent: Boolean = false
    ): List<Article>
}

enum class SortBy(val value: String) {
    POPULARITY("popularity"),
    PUBLISHED_AT("publishedAt"),
    RELEVANCY("relevancy")
}

enum class Country(val value: String) {
    AE("ae"), AR("ar"), AT("at"), AU("au"), BE("be"), BG("bg"), BR("br"), CA("ca"),
    CH("ch"), CN("cn"), CO("co"), CU("cu"), CZ("cz"), DE("de"), EG("eg"), FR("fr"),
    GB("gb"), GR("gr"), HK("hk"), HU("hu"), ID("id"), IE("ie"), IL("il"), IN("in"),
    IT("it"), JP("jp"), KR("kr"), LT("lt"), LV("lv"), MA("ma"), MX("mx"), MY("my"),
    NG("ng"), NL("nl"), NO("no"), NZ("nz"), PH("ph"), PL("pl"), PT("pt"), RO("ro"),
    RS("rs"), RU("ru"), SA("sa"), SE("se"), SG("sg"), SI("si"), SK("sk"), TH("th"),
    TR("tr"), TW("tw"), UA("ua"), US("us"), VE("ve"), ZA("za")
}

enum class Category(val value: String) {
    BUSINESS("business"),
    ENTERTAINMENT("entertainment"),
    GENERAL("general"),
    HEALTH("health"),
    SCIENCE("science"),
    SPORTS("sports"),
    TECHNOLOGY("technology");

    val displayName: String
        get() = value.replaceFirstChar { it.uppercase() }
}
